netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -585,6 +585,7 @@ private void sendGrpcFrame(ChannelHandlerContext ctx, SendGrpcFrameCommand cmd,
     }
     // Call the base class to write the HTTP/2 DATA frame.
     encoder().writeData(ctx, cmd.streamId(), cmd.content(), 0, cmd.endStream(), promise);
+    transportTracer.reportMessageSent(cmd.getNumMessages());
```

Comment:
Note that the message has not been sent here. It is simply queued. If the remote is applying push-back, it could be a long time until the message is sent (which is why the previous code was delaying until the future completed).

Reply:
Moved the incr back to the future callback.